File: connection/cluster/ClusterDatabaseManager.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.connection.cluster;

import com.vaticle.typedb.client.api.database.Database;
import com.vaticle.typedb.client.api.database.DatabaseManager;
import com.vaticle.typedb.client.common.exception.TypeDBClientException;
import com.vaticle.typedb.client.connection.TypeDBDatabaseManagerImpl;
import com.vaticle.typedb.common.collection.Pair;
import com.vaticle.typedb.protocol.ClusterDatabaseProto;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Client.CLUSTER_ALL_NODES_FAILED;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Client.CLUSTER_REPLICA_NOT_PRIMARY;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Client.DB_DOES_NOT_EXIST;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Cluster.DatabaseManager.allReq;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Cluster.DatabaseManager.getReq;
import static com.vaticle.typedb.common.collection.Collections.pair;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;

public class ClusterDatabaseManager implements DatabaseManager.Cluster {

    private static final Logger LOG = LoggerFactory.getLogger(ClusterDatabaseManager.class);

    private final Map<String, TypeDBDatabaseManagerImpl> databaseMgrs;
    private final ClusterClient client;

    public ClusterDatabaseManager(ClusterClient client) {
        this.client = client;
        this.databaseMgrs = client.clusterServerClients().entrySet().stream()
                .map(c -> pair(c.getKey(), c.getValue().databases()))
                .collect(toMap(Pair::first, Pair::second));
    }

    @Override
    public boolean contains(String name) {
        try {
            return failsafeTask(name, ((stub, dbMgr) -> dbMgr.contains(name))).runPrimaryReplica();
        } catch (TypeDBClientException e) {
            if (e.getErrorMessage().equals(DB_DOES_NOT_EXIST)) return false;
            else throw e;
        }
    }

    @Override
    public void create(String name) {
        ClusterClient.FailsafeTask<Void> failsafeTask = failsafeTask(name, (stub, dbMgr) -> {
            dbMgr.create(name);
            return null;
        });
        try {
            failsafeTask.runAnyReplica();
        } catch (TypeDBClientException e) {
            if (CLUSTER_REPLICA_NOT_PRIMARY.equals(e.getErrorMessage())) {
                failsafeTask.runPrimaryReplica();
            } else throw e;
        }
    }

    @Override
    public Database.Cluster get(String name) {
        return failsafeTask(name, (stub, dbMgr) -> {
            if (dbMgr.contains(name)) {
                ClusterDatabaseProto.ClusterDatabaseManager.Get.Res res = stub.databasesGet(getReq(name));
                return ClusterDatabase.of(res.getDatabase(), client);
            } else throw new TypeDBClientException(DB_DOES_NOT_EXIST, name);
        }).runPrimaryReplica();
    }

    @Override
    public List<Database.Cluster> all() {
        StringBuilder errors = new StringBuilder();
        for (String address : databaseMgrs.keySet()) {
            try {
                ClusterDatabaseProto.ClusterDatabaseManager.All.Res res = client.clusterServerClient(address)
                        .stub().databasesAll(allReq());
                return res.getDatabasesList().stream().map(db -> ClusterDatabase.of(db, client)).collect(toList());
            } catch (TypeDBClientException e) {
                errors.append("- ").append(address).append(": ").append(e).append("\n");
            }
        }
        throw new TypeDBClientException(CLUSTER_ALL_NODES_FAILED, errors.toString());
    }

    Map<String, TypeDBDatabaseManagerImpl> databaseMgrs() {
        return databaseMgrs;
    }

    private <RESULT> ClusterClient.FailsafeTask<RESULT> failsafeTask(String name, BiFunction<ClusterServerStub, TypeDBDatabaseManagerImpl, RESULT> task) {
        return client.createFailsafeTask(
                name,
                parameter -> task.apply(
                        parameter.client().stub(),
                        parameter.client().databases()
                )
        );
    }
}

File: connection/cluster/ClusterDatabaseManager.ts
/*
 * Copyright (C) 2021 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {ClusterServerStub} from "./ClusterServerStub";
import {ClusterDatabase} from "./ClusterDatabase";
import {ClusterClient} from "./ClusterClient";
import {FailsafeTask} from "./FailsafeTask";
import {TypeDBDatabaseManagerImpl} from "../TypeDBDatabaseManagerImpl";
import {Database} from "../../api/connection/database/Database";
import {DatabaseManager} from "../../api/connection/database/DatabaseManager";
import {RequestBuilder} from "../../common/rpc/RequestBuilder";
import {ErrorMessage} from "../../common/errors/ErrorMessage";
import {TypeDBClientError} from "../../common/errors/TypeDBClientError";
import {ClusterDatabaseManager as ClusterDatabaseManagerProto} from "typedb-protocol/cluster/cluster_database_pb";
import CLUSTER_ALL_NODES_FAILED = ErrorMessage.Client.CLUSTER_ALL_NODES_FAILED;
import CLUSTER_REPLICA_NOT_PRIMARY = ErrorMessage.Client.CLUSTER_REPLICA_NOT_PRIMARY;
import DB_DOES_NOT_EXIST = ErrorMessage.Client.DB_DOES_NOT_EXIST;

export class ClusterDatabaseManager implements DatabaseManager.Cluster {

    private readonly _databaseManagers: { [serverAddress: string]: TypeDBDatabaseManagerImpl };
    private readonly _client: ClusterClient;

    constructor(client: ClusterClient) {
        this._client = client;
        this._databaseManagers = Object.entries(this._client.clusterServerClients()).reduce((obj: { [address: string]: TypeDBDatabaseManagerImpl }, [addr, client]) => {
            obj[addr] = client.databases();
            return obj;
        }, {});
    }

    async contains(name: string): Promise<boolean> {
        return await this.failsafeTask(name, ((stub, dbMgr) => dbMgr.contains(name)));
    }

    async create(name: string): Promise<void> {
        return await this.failsafeTask(name, ((stub, dbMgr) => dbMgr.create(name)));
    }

    async get(name: string): Promise<Database.Cluster> {
        return await this.failsafeTask(name, (async (stub, dbMgr) => {
            if (await this.contains(name)) {
                const res: ClusterDatabaseManagerProto.Get.Res = await stub.databasesClusterGet(RequestBuilder.Cluster.DatabaseManager.getReq(name));
                return ClusterDatabase.of(res.getDatabase(), this._client);
            }
            throw new TypeDBClientError(DB_DOES_NOT_EXIST.message(name));
        }));
    }

    async all(): Promise<Database.Cluster[]> {
        let errors = "";
        for (const address of Object.keys(this._databaseManagers)) {
            try {
                const res = await this._client.stub(address).databasesClusterAll(RequestBuilder.Cluster.DatabaseManager.allReq());
                return res.getDatabasesList().map(db => ClusterDatabase.of(db, this._client));
            } catch (e) {
                errors += `- ${address}: ${e}\n`;
            }
        }
        throw new TypeDBClientError(CLUSTER_ALL_NODES_FAILED.message(errors));
    }

    databaseManagers(): { [address: string]: TypeDBDatabaseManagerImpl } {
        return this._databaseManagers;
    }

    rpcClient() {
        return this._client;
    }

    private async failsafeTask<T>(name: string, task: (stub: ClusterServerStub, dbMgr: DatabaseManager) => Promise<T>): Promise<T> {
        const failsafeTask = new DatabaseManagerFailsafeTask(this._client, name, task);
        try {
            return await failsafeTask.runAnyReplica();
        } catch (e) {
            if (e instanceof TypeDBClientError && CLUSTER_REPLICA_NOT_PRIMARY === e.errorMessage()) {
                return await failsafeTask.runPrimaryReplica();
            } else throw e;
        }
    }
}

class DatabaseManagerFailsafeTask<T> extends FailsafeTask<T> {

    private readonly _task: (stub: ClusterServerStub, dbMgr: DatabaseManager) => Promise<T>;

    constructor(client: ClusterClient, database: string, task: (stub: ClusterServerStub, dbMgr: DatabaseManager) => Promise<T>) {
        super(client, database);
        this._task = task;
    }

    async run(replica: Database.Replica): Promise<T> {
        return this._task(this.client.stub(replica.address()), this.client.clusterServerClient(replica.address()).databases());
    }
}
